File: TypeScript/day03/part2.ts
import { getPuzzleInput, createAlphabet } from "../utils";
const input = getPuzzleInput("day03");

export default function part2(input: string) {
  const arr = input.split("\n");
  const letters = createAlphabet();
  let groups: string[][] = [];
  let priorities = 0;

  while (arr.length > 0) {
    groups = [...groups, arr.splice(0, 3)];
  }

  groups.forEach((group) => {
    const common = new Set<string>();
    const [r1, r2, r3] = group;

    r1.split("").forEach((v) => r2.includes(v) && common.add(v));
    common.forEach((v) => {
      if (r3.includes(v)) priorities += letters.indexOf(v) + 1;
    });
  });

  return priorities;
}

File: Go/day03/part2.go
package main

import (
	"adventofcode/packages/utils"
	"strings"
)

func part2() int {
	content := utils.GetPuzzleInput("day03")
	groups := make([][]string, 0)
	alphabet := utils.CreateAlphabet()
	p := 0

	for i := 0; i < len(content); i += 3 {
		groups = append(groups, content[i:i+3])
	}

	for _, g := range groups {
		common := make(map[string]bool, 0)
		r1, r2, r3 := g[0], g[1], g[2]

		for _, v1 := range r1 {
			if strings.Contains(r2, string(v1)) {
				common[string(v1)] = true
			}
		}
		for v2 := range common {
			if strings.Contains(r3, v2) {
				for i, l := range alphabet {
					if v2 == l {
						p += i + 1
					}
				}
			}
		}
	}

	return p
}
